lib/lightning/workflows/snapshot.ex

```diff
@@ -0,0 +1,192 @@
+      field :source_trigger_id, :binary_id
+      field :target_job_id, :binary_id
+
+      field :condition_type, Ecto.Enum,
+        values: [:on_job_success, :on_job_failure, :always, :js_expression]
+
+      field :condition_expression, :string
+      field :condition_label, :string
+      field :enabled, :boolean
+
+      field :inserted_at, :utc_datetime
+      field :updated_at, :utc_datetime
+    end
+
+    timestamps(type: :utc_datetime_usec, updated_at: false)
+  end
+
+  def new(workflow) do
+    cast(%__MODULE__{}, workflow, [:name, :workflow_id])
+    |> cast_embed(:jobs, with: &job_changeset/2)
+    |> cast_embed(:triggers, with: &trigger_changeset/2)
+    |> cast_embed(:edges, with: &edge_changeset/2)
+  end
+
+  @job_fields Lightning.Workflows.Job.__schema__(:fields) -- [:workflow_id]
+  @trigger_fields Lightning.Workflows.Trigger.__schema__(:fields) --
+                    [:workflow_id]
+  @edge_fields Lightning.Workflows.Edge.__schema__(:fields) -- [:workflow_id]
+
+  defp job_changeset(schema, params) do
+    schema
+    |> cast(params, @job_fields)
+    |> validate_required([:id, :inserted_at, :updated_at])
+  end
+
+  defp trigger_changeset(schema, params) do
+    schema
+    |> cast(params, @trigger_fields)
+    |> validate_required([:id, :inserted_at, :updated_at])
+  end
+
+  defp edge_changeset(schema, params) do
+    schema
+    |> cast(params, @edge_fields)
+    |> validate_required([:id, :inserted_at, :updated_at])
+  end
+
+  @associations_to_include [:jobs, :triggers, :edges]
+
+  @spec build(Workflow.t()) :: Ecto.Changeset.t()
+  def build(%Workflow{} = workflow) do
+    workflow
+    |> Repo.preload([:jobs, :edges, :triggers])
+    |> Map.from_struct()
+    |> Enum.into(%{}, fn {field, value} ->
+      case field do
+        field when field in @associations_to_include ->
+          {field, Enum.map(value, &Map.from_struct/1)}
+
+        :name ->
+          {field, value}
+
+        :id ->
+          {:workflow_id, value}
+
+        _not_matching ->
+          {nil, nil}
+      end
+    end)
+    |> new()
+  end
+
+  @spec create(Workflow.t()) ::
+          {:ok, t()} | {:error, Ecto.Changeset.t()}
+  def create(%Workflow{} = workflow) do
+    build(workflow)
+    |> Repo.insert()
+  end
+
+  @spec get_all_for(Workflow.t()) :: [t()]
+  def get_all_for(%Workflow{} = workflow) do
+    from(s in __MODULE__, where: s.workflow_id == ^workflow.id)
+    |> Repo.all()
+  end
+
+  @doc """
+  Get the latest snapshot for a workflow, based on the inserted_at field.
+  """
+  @spec get_latest_for(Workflow.t()) :: t() | nil
+  def get_latest_for(%Workflow{} = workflow) do
+    get_latest_query(workflow)
+    |> Repo.one()
+  end
+
+  defp get_latest_query(workflow) do
+    from(s in __MODULE__,
+      where: s.workflow_id == ^workflow.id,
+      order_by: [desc: s.inserted_at],
+      limit: 1
+    )
+  end
+
+  @doc """
+  Get the latest snapshot for a workflow, or create one if it doesn't exist.
```

Comment:
I really like this. I'm assuming that it will be called when one of those "oh dang we really need a snapshot now" actions is performed:
1. manually initiate a github sync (must send _current_ version when button is pressed)
2. execute a run
3. something else maybe?

Reply:
Yep thats the case